sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetColumn.java

```diff
@@ -0,0 +1,239 @@
+    return result;
+  }
+
+  /**
+   * Assemble this column and calculate collection offsets recursively.
+   * This is a no-op for primitive columns.
+   */
+  void assemble() {
+    DataType type = columnInfo.sparkType();
+    if (type instanceof ArrayType || type instanceof MapType) {
+      for (ParquetColumn child : children) {
+        child.assemble();
+      }
+      calculateCollectionOffsets();
+    } else if (type instanceof StructType) {
+      for (ParquetColumn child : children) {
+        child.assemble();
+      }
+      calculateStructOffsets();
+    }
+  }
+
+  ParquetTypeInfo getColumnInfo() {
+    return this.columnInfo;
+  }
+
+  WritableColumnVector getValueVector() {
+    return this.vector;
+  }
+
+  WritableColumnVector getRepetitionLevelVector() {
+    return this.repetitionLevels;
+  }
+
+  WritableColumnVector getDefinitionLevelVector() {
+    return this.definitionLevels;
+  }
+
+  VectorizedColumnReader getColumnReader() {
+    return this.columnReader;
+  }
+
+  void setColumnReader(VectorizedColumnReader reader) {
+    if (!isPrimitive) {
+      throw new IllegalStateException("can't set reader for non-primitive column");
+    }
+    this.columnReader = reader;
+  }
+
+  private static void getLeavesHelper(ParquetColumn column, List<ParquetColumn> coll) {
+    if (column.isPrimitive) {
+      coll.add(column);
+    } else {
+      for (ParquetColumn childCol : column.children) {
+        getLeavesHelper(childCol, coll);
+      }
+    }
+  }
+
+  private void calculateCollectionOffsets() {
+    int maxDefinitionLevel = columnInfo.definitionLevel();
+    int maxElementRepetitionLevel = columnInfo.repetitionLevel();
+
+    // `i` is the index over all leaf elements of this array, while `offset` is the index over
+    // all top-level elements of this array.
+    for (int i = 0, rowId = 0, offset = 0; i < definitionLevels.getElementsAppended();
+         i = getNextCollectionStart(maxElementRepetitionLevel, i), rowId++) {
+      vector.reserve(rowId + 1);
+      int definitionLevel = definitionLevels.getInt(i);
+      if (definitionLevel == maxDefinitionLevel - 1) {
+        // the collection is null
+        vector.putNull(rowId);
+      } else if (definitionLevel == maxDefinitionLevel) {
+        // collection is defined but empty
+        vector.putNotNull(rowId);
+        vector.putArray(rowId, offset, 0);
+      } else {
+        // collection is defined and non-empty: find out how many top element there is till the
+        // start of the next array.
+        vector.putNotNull(rowId);
+        int length = getCollectionSize(maxElementRepetitionLevel, i + 1);
+        vector.putArray(rowId, offset, length);
+        offset += length;
+      }
+    }
+  }
+
+  private void calculateStructOffsets() {
+    int maxDefinitionLevel = columnInfo.definitionLevel();
+    vector.reserve(definitionLevels.getElementsAppended());
+    for (int i = 0, rowId = 0; i < definitionLevels.getElementsAppended(); i++, rowId++) {
```

Comment:
nit: definitionLevels.getElementsAppended() is called twice.

Reply:
BTW, isn't `rowId` as same as `i` in this loop?

Reply:
Oops yeah only one is needed.